TradeItIosTicketSDK2/EquitySymbolDataSource.swift

```diff
@@ -0,0 +1,121 @@
+import PromiseKit
+
+protocol SymbolDataSource: UITableViewDataSource {
+    var results: [TradeItSymbol] { get }
+
+    init(linkedBrokerAccount: TradeItLinkedBrokerAccount)
+
+    func searchByTextField(
+        textField: UITextField,
+        onSuccess: @escaping () -> Void,
+        onFailure: @escaping () -> Void
+    )
+}
+
+class CryptoSymbolDataSource: NSObject, SymbolDataSource {
+    var results: [TradeItSymbol] = []
+
+    private let linkedBrokerAccount: TradeItLinkedBrokerAccount
+    private var resultsPromise: Promise<[TradeItSymbol]>? = nil
+
+    required init(linkedBrokerAccount: TradeItLinkedBrokerAccount) {
+        self.linkedBrokerAccount = linkedBrokerAccount
+    }
+
+    // TODO: Can this just be promises all the way?
+    func searchByTextField(
+        textField: UITextField,
+        onSuccess: @escaping () -> Void,
+        onFailure: @escaping () -> Void
+    ) {
+        let queryText = textField.text ?? ""
+
+        getAllSymbolsPromise().then { (symbols) -> Void in
+            let topResults = symbols.filter { $0.symbol?.contains(queryText) ?? false }.prefix(20)
+            self.results = Array(topResults)
```

Comment:
Is `topResults` not already `[TradeItSymbol]`

Reply:
It's an array slice :<

Reply:
Ah -- so pointing to the same objects in memory? GTK.

Reply:
mm, nvm just confirmed my suspicions. This isn't doin what we think. Seems like swift array operations default to shallow copies (not sure why I was surprised by that). So, `Array(topResults)` is still going to point to the same location in memory.

Check out the swift repl and run the following commands:
```
> class A {
   var x = 5
}
> var a1 = [A(), A()]
> a1
[A] = 2 values {                                             // OUTPUT
  [0] = {
    a = 5
  }
  [1] = {
    a = 5
  }
}
> var a2 = Array(a1.prefix(1))
> a2[0].x = 6
> a1
[A] = 2 values {
  [0] = {
    a = 5
  }
  [1] = {
    a = 6
  }
}
```

Reply:
@oliver-was-here this is just converting `ArraySlice` in to an `Array`